woocommerce/woocommerce
# Created Repository Merge Tool

### All Submissions:


### Changes proposed in this Pull Request:

In the interest of being able to easily merge other repositories into this one, I have created this tool. Once executed it:

1. Install [`git-filter-repo`](https://github.com/woocommerce/woocommerce/pull/31974)
2. Clones the target repository into a temporary directory
3. Rewrites the history of that repository as if it always existed within a subdirectory here
4. Rewrites the commit messages to replace local links (like #208) with remote links (https://github.com/repo/pulls/208) so that they still work
5. **THIS IS DESTRUCTIVE AND CAN NOT BE REVERSED**: Merges the two histories together as if they were written in the same repository 

After the tool has finished executing you can use `git push` as you would expect. We can do this in a separate branch and then merge the history into `trunk`. Before considering this pull request ready to merge we need to investigate:

- Are the commit hashes of `woocommerce/woocommerce` maintained? They should be.
- What impact does this have on GitHub's references? Does everything still work?
- How are in-progress pull requests affected? Does rebase work correctly with the history manipulation we've done?

### How to test the changes in this Pull Request:

1. Check out this branch, do _not_ run the command yet, it will break your local branch
2. Run `pnpm install` to install package dependencies
2. Run `git checkout -b test/repository-merge` so you're on a copy of the PR's branch, giving you the tool, but not accidentally destroying everything in the process.
3. Run `./tools/monorepo-merge/bin/dev merge --help` and you can review the help page for the command.
4. Run `./tools/monorepo-merge/bin/dev merge woocommerce/woocommerce-admin plugins/woocommerce-admin`
5. Review the new `plugins/woocommerce-admin` folder in your local repository
6. Review [this PR that contains the result of the tool](https://github.com/woocommerce/woocommerce/pull/31975).

### Other information:




### Changelog entry

N/A

### FOR PR REVIEWER ONLY:
